src/client/parser.rs

```diff
@@ -38,42 +38,35 @@ impl HttpResponseParser {
     where
         T: IoStream,
     {
-        // if buf is empty parse_message will always return NotReady, let's avoid that
-        if buf.is_empty() {
-            match io.read_available(buf) {
-                Ok(Async::Ready((_, true))) => {
-                    return Err(HttpResponseParserError::Disconnect)
-                }
-                Ok(Async::Ready((_, false))) => (),
-                Ok(Async::NotReady) => return Ok(Async::NotReady),
-                Err(err) => return Err(HttpResponseParserError::Error(err.into())),
-            }
-        }
-
         loop {
-            match HttpResponseParser::parse_message(buf)
-                .map_err(HttpResponseParserError::Error)?
-            {
-                Async::Ready((msg, decoder)) => {
-                    self.decoder = decoder;
-                    return Ok(Async::Ready(msg));
-                }
-                Async::NotReady => {
-                    if buf.capacity() >= MAX_BUFFER_SIZE {
-                        return Err(HttpResponseParserError::Error(ParseError::TooLarge));
+            // Don't call parser until we have data to parse.
+            if !buf.is_empty() {
+                match HttpResponseParser::parse_message(buf)
+                    .map_err(HttpResponseParserError::Error)?
+                {
+                    Async::Ready((msg, decoder)) => {
+                        self.decoder = decoder;
+                        return Ok(Async::Ready(msg));
                     }
-                    match io.read_available(buf) {
-                        Ok(Async::Ready((_, true))) => {
-                            return Err(HttpResponseParserError::Disconnect)
-                        }
-                        Ok(Async::Ready((_, false))) => (),
-                        Ok(Async::NotReady) => return Ok(Async::NotReady),
-                        Err(err) => {
-                            return Err(HttpResponseParserError::Error(err.into()))
+                    Async::NotReady => {
+                        if buf.capacity() >= MAX_BUFFER_SIZE {
+                            return Err(HttpResponseParserError::Error(ParseError::TooLarge));
                         }
+                        // Parser needs more data.
                     }
                 }
             }
+            // Read some more data into the buffer for the parser.
+            match io.read_available(buf) {
```

Comment:
I feel like this kind of thing could very easily go into a busy loop that always reads zero bytes

Reply:
if `read_available()` isn't able to ready any data that will be handled by the `NotReady` branch which will return from the `poll()` method.

Reply:
Cool, I just tested it with our code and it seems to work fine.